bundle/src/main/java/com/adobe/acs/commons/packagegarbagecollector/PackageGarbageCollectionJob.java

```diff
@@ -118,120 +126,78 @@ public JobResult process(Job job) {
-                .max(Comparator.comparing(pkg -> new PackageDefinition(pkg).getLastUnpacked()));
-
-        if (lastInstalledPackageOptional.isPresent()) {
-            JcrPackage lastInstalledPackage = lastInstalledPackageOptional.get();
-            PackageDefinition lastInstalledPackageDefinition = new PackageDefinition(lastInstalledPackage);
-            PackageDefinition thisPackageDefinition = new PackageDefinition(jcrPackage);
-
-            // If it's not actually installed yet.
-            if (lastInstalledPackageDefinition.getLastUnpacked() == null) {
-                // This should never be here since this check is guarded by isInstalled() above.
-                return false;
-            }
+    private static final class UncheckedRepositoryException extends RuntimeException {
+        private static final long serialVersionUID = 8851421623772855854L;
 
-            return lastInstalledPackageDefinition.hasSamePid(thisPackageDefinition);
+        protected UncheckedRepositoryException(RepositoryException e) {
+            super(e);
         }
 
-        return false;
-    }
-
-    static class PackageDefinition {
-        JcrPackage jcrPackage;
-
-        public PackageDefinition(@Nonnull JcrPackage jcrPackage) {
-            this.jcrPackage = jcrPackage;
+        /**
+         * Returns the cause of this exception.
+         *
+         * @return  the {@code RepositoryException} which is the cause of this exception.
+         */
+        @Override
+        public RepositoryException getCause() {
+            return (RepositoryException) super.getCause();
         }
 
-        public Calendar getLastUnpacked() {
-            try {
-                JcrPackageDefinition definition = jcrPackage.getDefinition();
-                if (definition != null) {
-                    return definition.getLastUnpacked();
-                }
-                return null;
-            } catch (RepositoryException ex) {
-                return null;
-            }
-        }
+    }
 
-        public boolean isSameNameAndGroup(JcrPackage otherPackage) {
-            Optional<PackageId> otherPackageId = getPid(otherPackage);
-            Optional<PackageId> thisPackageId = getPid(jcrPackage);
-            if (otherPackageId.isPresent() && thisPackageId.isPresent()) {
-                return otherPackageId.get().getGroup().equals(thisPackageId.get().getGroup())
-                        && otherPackageId.get().getName().equals(thisPackageId.get().getName());
+    private boolean isLatestInstalled(JcrPackageDefinition referencePkgDefinition, Stream<JcrPackage> installedPackages) throws RepositoryException {
+        try {
+            Optional<@Nullable JcrPackageDefinition> lastInstalledPckDefinitionOptional = installedPackages
+               .map(p -> {
+                    try {
+                        return p.getDefinition();
+                    } catch (RepositoryException e) {
+                        throw new UncheckedRepositoryException(e);
```

Comment:
Does this exception contain enough metadata to identify the package where the exception happened? Or should we add an additional message to it?

Reply:
I now enrich the relevant exception with the package path